aten/src/ATen/native/cuda/SummaryOps.cu

```diff
@@ -352,11 +353,100 @@ Tensor _histc_cuda_template(
       "] is not finite");
 #endif
   TORCH_CHECK(minvalue < maxvalue, "max must be larger than min");
-
   auto ret = cuda::CUDA_tensor_histogram<input_t, input_t, false>(
-    output, self, Tensor(), nbins, minvalue, maxvalue);
+      output, self, Tensor(), nbins, minvalue, maxvalue);
   return output;
 }
+
+///////////////// histogram /////////////////
+template <typename input_t, typename weights_t>
+std::tuple<Tensor,Tensor> _histogram_cuda_template_uniform_bins(
+    const Tensor& self,
+    int64_t nbins,
+    const Tensor& weights,
+    c10::optional<ArrayRef<double>> range,
+    bool density) {
+  input_t minvalue;
+  input_t maxvalue;
+  if (range.has_value()) {
+    // If range is defined, max must be larger than min.
+    TORCH_CHECK(
+        range.value()[0] < range.value()[1], "max must be larger than min");
+    minvalue = static_cast<input_t>(range.value()[0]);
+    maxvalue = static_cast<input_t>(range.value()[1]);
+  } else {
+    minvalue = *self.min().cpu().data_ptr<input_t>();
+    maxvalue = *self.max().cpu().data_ptr<input_t>();
+    // This is done to avoid divide by zero if input min is equal to input max.
+    // In this case computing the histogram can also be skipped altogether, as
+    // it's equal to the sum of weights in the middle bin, and zero everywhere
+    // else.
+    if (minvalue == maxvalue) {
+      minvalue -= 1;
+      maxvalue += 1;
+    }
+  }
+
+  TORCH_CHECK(nbins > 0, "bins must be > 0");
+#ifndef __HIP_PLATFORM_HCC__
```

Comment:
This code is still running on the CPU so it can share the same preamble as the CPU version. The only difference in the CUDA version is the call to CUDA_tensor_histogram. As mentioned previously, you could probably combine the CPU and GPU-specific versions of this function and have a couple checks for device type that do the (tiny) device-specific parts of the computation.